xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -188,51 +185,68 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
                       .withDescription("No traffic director provided by bootstrap")));
           return;
         }
-        // Currently we only support using the first server from bootstrap.
-        ServerInfo serverInfo = serverList.get(0);
-        channel = initLbChannel(
-            lookasideLbHelper, serverInfo.getServerUri(),
-            serverInfo.getChannelCredentials());
-        xdsClientPool = new RefCountedXdsClientObjectPool(new XdsClientFactory() {
+        XdsClientFactory xdsClientFactory = new XdsClientFactory() {
           @Override
           XdsClient createXdsClient() {
-            // TODO(zdapeng): Replace XdsComms2 with XdsClientImpl.
-            return new XdsComms2(
-                channel, lookasideLbHelper, new ExponentialBackoffPolicy.Provider(),
-                GrpcUtil.STOPWATCH_SUPPLIER, bootstrapInfo.getNode());
+            return
+                new XdsClientImpl(
+                    serverList,
+                    channelFactory,
+                    node,
+                    lookasideLbHelper.getSynchronizationContext(),
+                    lookasideLbHelper.getScheduledExecutorService(),
+                    new ExponentialBackoffPolicy.Provider(),
+                    GrpcUtil.STOPWATCH_SUPPLIER);
           }
-        });
+        };
+        xdsClientPool = new RefCountedXdsClientObjectPool(xdsClientFactory);
       }
       xdsClient = xdsClientPool.getObject();
     }
 
+    // The edsServiceName field is null in legacy gRPC client with EDS: use target authority for
+    // querying endpoints, but in the future we expect this to be explicitly given by EDS config.
+    // We assume if edsServiceName is null, it will always be null in later resolver updates;
+    // and if edsServiceName is not null, it will always be not null.
+    String clusterServiceName = newXdsConfig.edsServiceName;
+    if (clusterServiceName == null) {
+      clusterServiceName = lookasideLbHelper.getAuthority();
+    }
+    if (clusterName == null) {
+      // TODO(zdapeng): Use the correct cluster name. Currently load reporting will be broken if
+      //     edsServiceName is changed because we are using edsServiceName for the cluster name.
+      clusterName = clusterServiceName;
+    }
+
+    boolean shouldReportStats = newXdsConfig.lrsServerName != null;
+    if (shouldReportStats && !isReportingStats()) {
+      // Start load reporting. This may be a restarting after previously stopping the load
+      // reporting, so need to re-add all the pre-existing loadStatsStores to the new
+      // loadReportClient.
+      loadReportClient = xdsClient.reportClientStats(clusterName, newXdsConfig.lrsServerName);
+      for (Map.Entry<String, LoadStatsStore> entry : loadStatsStoreMap.entrySet()) {
```

Comment:
A `LoadStatsStore` is created during the construction of a `ClusterEndpointsBalancer` instance, which happens later. So here you won't add any `LoadStatsStore` to the load report client.

Seems the `LoadReportClient` and the `XdsClient`'s API for start/stop load reporting is a bit awkward to use here. What about we change the specification a little bit? We expose `start()/stop()` methods in `LoadReportClient` and they can be called multiple times. That is, in `LookasideLb` we call `xdsClient.reportClientStats(...)` at most once. `LookasideLb` holds the load report client instance. If the config changes to no longer do load reporting, just call the load report client's `stop()`. If config changes again to do load reporting again, just call `start()`. In this way, we are not changing the load report client instance and previously added `LoadStatsStore` do not need to be re-added after resume load reporting. What do you think?